src/charm.py

```diff
@@ -38,77 +38,87 @@ def __init__(self, *args: typing.Any):
+        self, event: typing.Union[ops.ConfigChangedEvent, ops.UpgradeCharmEvent]
+    ) -> None:
+        """Register the agent to server from configuration values.
 
         Args:
-            event: The base hook event.
+            event: The event fired on config changed or upgrade charm.
         """
-        container = self._jenkins_agent_container
+        container = self.unit.get_container(self.state.jenkins_agent_service_name)
         if not container.can_connect():
             logger.warning("Jenkins agent container not yet ready. Deferring.")
             event.defer()
             return
 
         if (
             not self.state.jenkins_config
-            and not self.model.get_relation(AGENT_RELATION)
             and not self.model.get_relation(SLAVE_RELATION)
+            and not self.model.get_relation(AGENT_RELATION)
         ):
             self.model.unit.status = ops.BlockedStatus("Waiting for config/relation.")
             return
 
         if not self.state.jenkins_config:
+            if self.model.get_relation(SLAVE_RELATION):
+                self.model.unit.status = ops.BlockedStatus(
+                    "Please remove and re-relate slave relation."
+                )
+                return
+            # Support fallback relation to AGENT_RELATION.
+            self.model.unit.status = ops.BlockedStatus(
+                "Please remove and re-relate slave relation."
+            )
             return
 
         try:
             server.download_jenkins_agent(
                 server_url=self.state.jenkins_config.server_url,
-                connectable_container=container,
+                container=container,
             )
         except server.AgentJarDownloadError as exc:
             logger.error("Failed to download Agent JAR executable, %s", exc)
             self.model.unit.status = ops.BlockedStatus("Failed to download Agent JAR executable.")
             return
 
-        valid_agent_token: typing.Optional[typing.Tuple[str, str]] = None
-        for agent_name, agent_token in zip(
-            self.state.jenkins_config.agent_names, self.state.jenkins_config.agent_tokens
-        ):
-            logger.warning("[%s]Validating %s", self.state.agent_meta.name, agent_name)
-            if not server.validate_credentials(
-                agent_name=agent_name,
-                credentials=server.Credentials(
-                    address=self.state.jenkins_config.server_url, secret=agent_token
-                ),
-                connectable_container=container,
-            ):
-                logger.error(
-                    "[%s]agent %s validation failed.", self.state.agent_meta.name, agent_name
-                )
-                continue
-            valid_agent_token = (agent_name, agent_token)
-
+        valid_agent_token = server.find_valid_credentials(
+            agent_name_token_pairs=zip(
+                self.state.jenkins_config.agent_names, self.state.jenkins_config.agent_tokens
```

Comment:
This is for backwards compatibility, where agent names may not necessarily match the `agent_name` / `agent_token` pair. For instance, a Jenkins agent can be registered through the UI on the server and those custom created credentials could be used to register the agents through configuration values.

Reply:
I'm not quite sure if passing in the state would reduce coupling. My thought process is that the function should be generic within the context of it's own and be more standalone and for charm specific things we have a transforming function to map the data to the exposed generic functions, hence passing in the `agent_name_token_pairs` rather than the state.

Reply:
It is a question of where the backwards compatibility logic should be done. It is more business logic and hence not really appropriate for `charm.py`. There could be another function that takes care of the backwards compatibility which calls the current function

Reply:
I'm not sure I fully understand what you mean.
I think the business logic in this case is: Find the credential pair that can be used (out of all provided since we don't know which unit should be using what pair). 
This business logic is encapsulated in `server.find_valid_credentials` which takes in `agent_name_token_pairs` as an argument.
Would you mean that the `zip(agent_names, agent_tokens)` is the business logic? In that case, I think it's only mutating what is in the state to be appropriate for the function to receive, which I don't think is quite business logic. I think it is better to receive an iterable of agent name and agent token pairs rather than state because I think it is better to keep the business logic functions decoupled from any other paradigms within the codebase.

Reply:
Zipped in state.py! Thank you for the clarification